core/src/main/java/org/geysermc/geyser/session/cache/SkullCache.java

```diff
@@ -0,0 +1,210 @@
+
+    private final List<Skull> inRangeSkulls = new ArrayList<>();
+
+    private final Deque<SkullPlayerEntity> unusedSkullEntities = new ArrayDeque<>();
+    private int totalSkullEntities = 0;
+
+    private final GeyserSession session;
+
+    private Vector3i lastPlayerPosition;
+
+    private long lastCleanup = System.currentTimeMillis();
+
+    public SkullCache(GeyserSession session) {
+        this.session = session;
+        this.maxVisibleSkulls = session.getGeyser().getConfig().getMaxVisibleCustomSkulls();
+        this.cullingEnabled = this.maxVisibleSkulls != -1;
+
+        // Normal skulls are not rendered beyond 64 blocks
+        int distance = Math.min(session.getGeyser().getConfig().getCustomSkullRenderDistance(), 64);
+        this.skullRenderDistanceSquared = distance * distance;
+    }
+
+    public void putSkull(Vector3i position, GameProfile profile, int blockState) {
+        Skull skull = skulls.computeIfAbsent(position, Skull::new);
+        skull.profile = profile;
+        skull.blockState = blockState;
+
+        if (skull.entity != null) {
+            skull.entity.updateSkull(skull);
+        } else {
+            if (!cullingEnabled) {
+                assignSkullEntity(skull);
+                return;
+            }
+            if (lastPlayerPosition == null) {
+                return;
+            }
+            skull.distanceSquared = position.distanceSquared(lastPlayerPosition);
+            if (skull.distanceSquared < skullRenderDistanceSquared) {
+                // Keep list in order
+                int i = Collections.binarySearch(inRangeSkulls, skull, Comparator.comparingInt(Skull::getDistanceSquared));
+                if (i < 0) { // skull.distanceSquared is a new distance value
+                    i = -i - 1;
+                }
+                inRangeSkulls.add(i, skull);
+
+                if (i < maxVisibleSkulls) {
+                    // Reassign entity from the farthest skull to this one
+                    if (inRangeSkulls.size() > maxVisibleSkulls) {
+                        freeSkullEntity(inRangeSkulls.get(maxVisibleSkulls));
+                    }
+                    assignSkullEntity(skull);
+                }
+            }
+        }
+    }
+
+    public void removeSkull(Vector3i position) {
+        Skull skull = skulls.remove(position);
+        if (skull != null && skull.entity != null) {
+            freeSkullEntity(skull);
+
+            if (cullingEnabled) {
+                inRangeSkulls.remove(skull);
+                if (inRangeSkulls.size() >= maxVisibleSkulls) {
+                    // Reassign entity to the closest skull without an entity
+                    assignSkullEntity(inRangeSkulls.get(maxVisibleSkulls - 1));
+                }
+            }
+        }
+    }
+
+    public void updateVisibleSkulls() {
+        if (cullingEnabled) {
+            Vector3i playerPosition = session.getPlayerEntity().getPosition().toInt();
```

Comment:
Could you replace this with a Vector3f just so we don't have to create a new Vector3i instance every time the player moves? Looks like you could have lastPlayerPosition do that too, and when you call distanceSquared on the skull position you can just use raw numbers? Worst case just use toInt() after the distance squared check on the next line?